jupyterhub/apihandlers/auth.py

```diff
@@ -144,35 +144,89 @@ def make_absolute_redirect_uri(self, uri):
             ._replace(query=urlencode(query_list))
         )
 
+    def add_credentials(self, credentials=None):
+        """Add oauth credentials
 
-class OAuthAuthorizeHandler(OAuthHandler):
-    """Implement OAuth provider handlers
+        Adds user, session_id, client to oauth credentials
+        """
+        if credentials is None:
+            credentials = {}
+        else:
+            credentials = credentials.copy()
+
+        session_id = self.get_session_cookie()
+        if session_id is None:
+            session_id = self.set_session_cookie()
+
+        user = self.get_current_user()
 
-    OAuth2Handler sets `self.provider` in initialize,
-    but we are already passing the Provider object via settings.
-    """
+        # Extra credentials we need in the validator
+        credentials.update({
+            'user': user,
+            'handler': self,
+            'session_id': session_id,
+        })
+        return credentials
+
+    def send_oauth_response(self, headers, body, status):
+        """Send oauth response
+
+        headers, body, status  are returned by provider method.
```

Comment:
Perhaps add something like:

This method prepares headers, body, status for an oauth response.
The provider method returns headers, body, status in its response.

Reply:
Done, thanks!